haystack/document_store/milvus.py

```diff
@@ -0,0 +1,420 @@
+
+        status, ok = self.milvus_server.has_collection(collection_name=index)
+        if not ok:
+            collection_param = {
+                'collection_name': index,
+                'dimension': self.vector_dim,
+                'index_file_size': self.index_file_size,
+                'metric_type': self.metric_type
+            }
+
+            status = self.milvus_server.create_collection(collection_param)
+            if status.code != Status.SUCCESS:
+                raise RuntimeError(f'Collection creation on Milvus server failed: {status}')
+
+            status = self.milvus_server.create_index(index, self.index_type, index_param)
+            if status.code != Status.SUCCESS:
+                raise RuntimeError(f'Index creation on Milvus server failed: {status}')
+
+    def _create_document_field_map(self) -> Dict:
+        return {
+            self.index: self.embedding_field,
+        }
+
+    def write_documents(
+            self, documents: Union[List[dict], List[Document]], index: Optional[str] = None, batch_size: int = 10_000
+    ):
+        """
+        Add new documents to the DocumentStore.
+
+        :param documents: List of `Dicts` or List of `Documents`. If they already contain the embeddings, we'll index
+                                  them right away in FAISS. If not, you can later call update_embeddings() to create & index them.
+        :param index: (SQL) index name for storing the docs and metadata
+        :param batch_size: When working with large number of documents, batching can help reduce memory footprint.
+        :return:
+        """
+        index = index or self.index
+        self._create_collection_and_index_if_not_exist(index)
+        field_map = self._create_document_field_map()
+        document_objects = [Document.from_dict(d, field_map=field_map) if isinstance(d, dict) else d for d in documents]
+
+        add_vectors = False if document_objects[0].embedding is None else True
+
+        for i in range(0, len(document_objects), batch_size):
+            vector_ids = []
+            if add_vectors:
+                doc_ids = []
+                embeddings = []
+                for doc in document_objects[i: i + batch_size]:
+                    doc_ids.append(doc.id)
+                    if isinstance(doc.embedding, np.ndarray):
+                        embeddings.append(doc.embedding.tolist())
+                    elif isinstance(doc.embedding, list):
+                        embeddings.append(doc.embedding)
+                    else:
+                        raise AttributeError(f'Format of supplied document embedding {type(doc.embedding)} is not '
+                                             f'supported. Please use list or numpy.ndarray')
+
+                if self.update_existing_documents:
+                    existing_docs = super().get_documents_by_id(ids=doc_ids, index=index)
+                    self._delete_vector_ids_from_milvus(documents=existing_docs, index=index)
```

Comment:
When we update an existing document, what happens to the previously indexed vector? Does it gets replaced or is it still indexed in Milvus?

Reply:
it replaced it. See this test cover this https://github.com/deepset-ai/haystack/blob/d57b9e7beb3cf29dc64eb129c247c872ec55dd7d/test/test_faiss_and_milvus.py#L89

Reply:
Please suggest any other test case which we might need to cover as well